File: src/zros_cpp/impl/publisher_impl.cc
//
// Created by pairs on 7/25/18.
//

#include <sspdlog/sspdlog.h>
#include "publisher_impl.h"
#include "zeromq_help.h"

namespace zros {

    ZmqPubStub::ZmqPubStub(const std::string& topic):PubStub(topic) {
        context_ = std::make_shared<zmq::context_t>(1);
        pub_stub_ = std::make_shared<zmq::socket_t>(*context_.get(), ZMQ_PUB);
        pub_stub_->bind("tcp://*:*");
        const size_t address_size = 32;
        char address[address_size];
        auto rc = zmq_getsockopt(*pub_stub_.get(), ZMQ_LAST_ENDPOINT, address, (size_t*)&address_size);
        assert(rc == 0);
        this->set_address(address);
    }

    bool ZmqPubStub::publish(const std::string& message) {
        std::lock_guard<std::mutex> lk(mtx_);
        return zeromq::ZmqHelp::s_send(*pub_stub_.get(), message);
    }

    //////////////////// publishersImpl///////////////////////////
    PublishersImpl::PublishersImpl() {
    }

    bool PublishersImpl::publish(const std::string &topic, const std::string &message) {
        return false;
    }

    void PublishersImpl::registerPublisher(const std::shared_ptr<IPublisher> publisher) {
        std::lock_guard<std::mutex> lk(pubs_mutex_);
        if (publishers_.find(publisher->get_topic()) != publishers_.end()) {
            SSPD_LOG_WARNING << "the publisher " << publisher->get_topic() << " already register";
            publisher->set_address(publishers_[publisher->get_topic()]->get_address());
            return;
        }
        auto address = _create_publisher(publisher->get_topic());
        publisher->set_address(address);
    }

    // return pub address
    const std::string &PublishersImpl::_create_publisher(const std::string &topic) {
        publishers_[topic] = std::make_shared<ZmqPubStub>(topic);
        const std::string& address = publishers_[topic]->get_address();
        SSPD_LOG_INFO << "registerPublisher: " << topic << " on address " << address << " success";
        return address;
    }
}   // namespace zros

File: src/zros_python/impl/publisher_impl.py
# -*- coding: utf-8 -*-
""" impl publisher management

data: 2018-08-02
author: pairs
"""
import abc
import logging
import threading
import zmq

logger = logging.getLogger(__name__)


class PubStub(object):
    __metaclass__ = abc.ABCMeta

    def __init__(self, topic):
        self._topic = topic
        self._address = None

    @abc.abstractmethod
    def publish(self, message):
        pass

    def set_address(self, address):
        self._address = address

    def get_address(self):
        return self._address


class ZmqPubStub(PubStub):
    """

    """
    def __init__(self, topic):
        super(ZmqPubStub, self).__init__(topic)
        self._mtx = threading.Lock()
        self._context = zmq.Context()
        self._pub_stub = self._context.socket(zmq.PUB)
        port_selected = self._pub_stub.bind_to_random_port('tcp://*', min_port=5000, max_port=65535, max_tries=100)
        self.set_address(u'tcp://localhost:' + str(port_selected))
        logger.info(u'topic %s bind on address %s', topic, self.get_address())

    def publish(self, message):
        with self._mtx:
            self._pub_stub.send_string(message)


class PublishersImpl(object):
    """

    """
    def __init__(self):
        self._pub_mutex = threading.Lock()
        self._publishers = {}

    def register_publisher(self, publisher):
        topic = publisher.get_topic()
        with self._pub_mutex:
            if topic in self._publishers:
                logger.warning(u'the publisher %s already register', topic)
                return
            address = self._create_publisher(topic)
            publisher.set_address(address)

    def publish(self, topic, message):
        with self._pub_mutex:
            if topic in self._publishers:
                self._publishers[topic].publish(message)
            else:
                logger.warning(u'please check publisher %s if exist', topic)

    def _create_publisher(self, topic):
        self._publishers[topic] = ZmqPubStub(topic)
        address = self._publishers[topic].get_address()
        logger.info(u'register publisher: %s on address success', address)
        return address
